src/subscriptions.rs

```diff
@@ -0,0 +1,148 @@
+        }
+    }
+
+    /// Wait for a block of packets to be filled and then send it through the channel.
+    /// Check the min number necessary to reconstruct the data set
+    /// randomly ask a peer or parent for the missing packets in random order
+    pub fn receiver(&self, r: streamer::Receiver, s: streamer::Sender)-> Result<()> {
+        let mut packets = self.window_peek();
+        let required = (* 100) / FANOUT;
+        loop {
+            packets.fill_packets(s)?;
+            let t = (FANOUT - CODES) * packets.len();
+            let min_needed = (t + (FANOUT / 2)) / FANOUT;
+            if packets.num_valid() <= min_needed {
+                break;
+            }
+            let mut num = packets.num_valid() - min_needed;
+            let c = self.peers() + self.parents();
+            for ix in shuffle(0..packets.len()) {
+                if num == 0 {
+                    break;
+                }
+                let p = packets[ix];
+                if p.valid() {
+                    continue;
+                }
+                let r = c[random(0, c.len())];
+                s.send_to(r, p.request());
+                num--;
+            }
+        }
+        let done = self.window_pop()?;
+        s.send(done);
+    }
+    /// Assume all the peers are sending to all the children
+    /// Broadcast to as many children as you can, avoid sending duplicates
+    /// p is parent, c is child
+    /// p1, p2
+    /// c1, c2, c3, c4
+    /// p1 sends packet: 1,5; 2,6;      3;      4
+    /// p2 sends packet: 4;   1,5;    2,6;      3
+    /// c1 sends:        _;     _;    1,5;    1,5
+    /// c2 sends:      2,6;     _;      _;    2,6
+    /// c3 sends:        3;     3;      _;      _
+    /// c4 sends:        _;     4;      4;      _  
+    pub fn rebroadcast(&self, sock: UdpSocket) -> Result<()> {
+        let packets = self.window_peak()?;
+        let me = self.my_peer_index();
+        let peers = self.peers();
+        let parents_len = self.parents().len();
+        for p in packets.iter_mut() {
+            if p.transmitted {
+                continue;
+            }
+            if !p.valid {
+                continue;
+            }
+            if p.index() % peers.len() == me {
+                let end = (me + parents_len) % peers.len();
+                if end < me {
+                    p.send_to(peers[end .. me])?;
+                } else {
+                    p.send_to(peers[0 .. me])?;
+                    p.send_to(peers[end .. ])?;
+                }
+            }
+            p.transmitted = true;
+        }
+    }
+    /// Wait for requests form peers, then respond to them
+    pub fn peer_server(&self, s: UdpSocket, request: Reciever<SharedPacketData>) -> Result<()> {
+        requests.recv(s)?;
+        for r in requests.iter() {
+            let req = deserialize(r.data[0..r.len]);
+            let packet = self.window_lookup(req);
+            packet.send_to(r.get_addr());
+        }
+    }
+}
```

Comment:
This module needs lots and lots of tests, especially for your rebroadcast function